### PR TITLE
fix: support full share attachments to fix facebook buttons

### LINKED ISSUE #402: Message payload doesn't contain buttons data when there is only one card
Thanks for a great library!

I'm trying on version 1.2.0, to fetch buttons from an attachment that I receive from a bot.  In the attachment (see below) there are 3 buttons but I can only get the title "search engines"

This is partial payload from this library:
```javascript
{ ...
  attachments:
   [ { ...
       type: 'share',
       source: 'Page - testing',
       title: 'search engines',
   } ],
  ...
}
```

This is the attachment:
image

When I'm sniffing facebook chat network I can see that the buttons are part of `call_to_actions` property under an `attachments[0]/mercury/share/target/items[0]` - item with the property `name: "search engines"`

When a bot sends me more than one attachment ("carousel"), I can see the subattachments with relevant details (including buttons)

I'm using the library https://github.com/bluejamesbond/FacebookMessengerBot.js and one of its examples as the bot

### PR DESCRIPTION
**Pull request by noamshemesh**
fix #402 

Not directly related: I think we should break `listen` to several functions to be able to test it better.

### CODE REVIEW

**Comment by Schmavery** [comment]
Great, thanks for fixing this.  Is there a reason why we shouldn't check if it's specifically the type of event we're trying to parse?
```
if (v.delta.attachments[i].mercury.attach_type == 'share') {
```

I agree that these switch statements have gotten a little out of hand and should probably be carefully refactored rather than continuing to be patched (turns out one of those is easier to do than the other, which explains the current status).  One first step might be to pull out this inlined resolveAttachmentUrl function.  Might help make it more readable.

I'm not sure I agree with your testing methodology here.  I understand that it's normal to freeze some input fixtures to have a deterministic and fast test, but I'm worried that in so doing, we fail to test the one thing that is most likely to break.  9 times out of 10, when something breaks, it is facebook sending a differently formatted response.  If we're testing against a frozen version, we will miss this bug.

File: src/listen.js [hunk 1]
```diff
@@ -146,6 +146,8 @@ module.exports = function(defaultFuncs, api, ctx) {
                         if (!err) v.delta.attachments[i].mercury.metadata.url = url;
                         return resolveAttachmentUrl(i + 1);
                       });
+                    } else {
+                      return resolveAttachmentUrl(i + 1);
```
Comment (Schmavery) [inline src/listen.js]: Nice catch

File: utils.js [hunk 2]
```diff
@@ -323,6 +323,8 @@ function _formatAttachment(attachment1, attachment2) {
         source: attachment1.share.source,
         title: attachment1.share.title,
         facebookUrl: attachment1.share.uri,
+        target: attachment1.share.target,
+        styleList: attachment1.share.style_list,
```
Comment (Schmavery) [inline utils.js]: Can these be documented?

**Comment by Schmavery** [comment]
Is there any change in behaviour after this change for regular attachments?

File: utils.js (hunk 2 above)
Comment (noamshemesh) [inline utils.js]: do you have any documentation example? I'm not sure what you expect :)

**Comment by noamshemesh** [comment]
Shouldn't affect other attachments besides that before adding the `else` block it didn't parse an attachment unless it was a photo (as you've already noticed)

**Comment by noamshemesh** [comment]
Regarding the testing methodology, I totally agree that we must have end to end tests that validate facebook API. 
But I think that we should have more unit tests. The advantages are:
1. we can write them fast
2. they run much much faster
3. we can test internal logic
And 3 is directly related to the reason I had to add this as a unit test. As you mentioned I can't send a button with this library to test the result.

What do you mean by "Does this new code only run when it sees a button"? Can you comment on the specific line?

**Comment by erdnaxeli** [comment]
Does this fix #415 too?

File: utils.js (hunk 2 above)
Comment (Schmavery) [inline utils.js]: https://github.com/Schmavery/facebook-chat-api/blob/master/DOCS.md#apilistencallback

**Comment by noamshemesh** [comment]
@erdnaxeli Yes, I've noticed the same problem and fixed it here: https://github.com/Schmavery/facebook-chat-api/pull/416/files/3c3481129fa13746a5a86fb33f597df4d901ea14#diff-f55ddb040d37ebcc409e74ccbdbdf1a2R150

**Comment by Schmavery** [comment]
Awesome, thanks @noamshemesh, I'll merge as soon as you add the new fields to the docs.

**Comment by noamshemesh** [comment]
Done, sorry for the late response.

**Comment by Schmavery** [comment]
Thanks :)

**Pull request merged**